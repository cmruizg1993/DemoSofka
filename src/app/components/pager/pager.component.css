.pages>ul{
    display: inline;
}
.pages>ul>li{
    display: inline;
    margin-left: 1rem;
    cursor: pointer;
}
.pages>ul>li>span{
    font-size: 1rem;
    color: var(--blue-600);
}
.pages>ul>li>span:hover{
    color: var(--blue-700);
    text-decoration: underline;
}
/*
.pages>ul>li>span.active{
    font-weight: 800;
}
*/