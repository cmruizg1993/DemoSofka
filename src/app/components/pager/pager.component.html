<div class="row p-2">
    <div class="col">
        <p class="fs-2 fw-800 fs-2 fw-800">Mostrando {{pagerProperties.pageSize}} de {{dataLength}}&emsp; Resultados</p>
    </div>
    <div class="col d-flex justify-content-end">
        <div class="pages">
            <ul>
                <li  *ngFor="let page of pages" (click)="updatePage(page)">
                    <span [class.active]="page.isActive">{{page.number}}</span>
                </li>
            </ul>
            <p class="text-primary fs-2 fw-800">Página actual: {{pagerProperties.page}}</p>
        </div>
    </div>
    <div class="col d-flex justify-content-end">
        <app-select 
        [data]="[{label: '5', value: 5, isSelected: true},{label: '10', value: 10},{label: '20', value: 20}]"
        label="Items por pág."
        (modelUpdate)="updatePageSize($event)">
        </app-select>
    </div>
</div>

