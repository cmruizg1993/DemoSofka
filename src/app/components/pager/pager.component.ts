import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'app-pager',
  templateUrl: './pager.component.html',
  styleUrl: './pager.component.css'
})
export class PagerComponent {
  @Output() pagerUpdate =  new EventEmitter<PagerProperties>();
  @Input() dataLength = 0;
  
  pagerProperties = new PagerProperties();
  
  pages: Page[] = [];
  ngOnInit(): void {
    //Called after the constructor, initializing input properties, and the first call to ngOnChanges.
    //Add 'implements OnInit' to the class.
    
  }
  ngAfterContentChecked(): void {
    //Called after every check of the component's or directive's content.
    //Add 'implements AfterContentChecked' to the class.
    if(this.dataLength > 0){
      const numberOfPages = Math.ceil(this.dataLength/this.pagerProperties.pageSize);
      this.pages = Array(numberOfPages).fill(null).map((x,i)=>{
        const page = new Page();
        page.number = i+1;
        page.isActive = i == 0 ;
        return page;
      });
      //console.log(this.pages)
    }
  }
  updatePage(page: Page){
    this.pagerProperties.page = page.number;
    this.updatePager();
  }
  updatePageSize(event: any){
    this.pagerProperties.page = 1;
    this.pagerProperties.pageSize = event.value;
    this.updatePager();
  }
  updatePager(){
    this.pagerUpdate.emit(this.pagerProperties);
  }
  
}
class PagerProperties{
  page: number = 1;
  pageSize = 5;
}
class Page{
  number: number = 0;
  isActive: boolean = false;
}
