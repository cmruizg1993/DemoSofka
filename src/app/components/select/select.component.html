<label *ngIf="label">
    {{label}}
</label>
<select [(ngModel)]="model">
    <option [ngValue]="item" *ngFor="let item of data">{{item.label}}</option>
</select>
