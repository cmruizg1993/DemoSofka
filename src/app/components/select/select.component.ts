import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'app-select',
  templateUrl: './select.component.html',
  styleUrl: './select.component.css'
})
export class SelectComponent {
  @Input() data: any[] = [];
  @Input() label = '';
  default: any;
  @Output() modelUpdate =  new EventEmitter<any>();
  private privateModel: any;

  ngOnInit(): void {
    //Called after the constructor, initializing input properties, and the first call to ngOnChanges.
    //Add 'implements OnInit' to the class.
    this.default = this.data.find( d => d.isSelected );
    if(this.default)
        this.model = this.default;
  }
  set model(newModel: any){
    this.privateModel = newModel;
    this.emitUpdate();
  }
  get model(){
    return this.privateModel;
  }
  
  emitUpdate(){
    //console.log(this.model)
    this.modelUpdate.emit(this.model);
  }
}
