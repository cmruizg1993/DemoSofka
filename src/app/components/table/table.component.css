
table{
    border-spacing: 0;
    overflow-x: scroll;
    width: 100%;
}
thead{
    background-color: var(--darkGrayishBlue50);
}
tr:hover{
    background-color: var(--darkGrayishBlue200);
}
th, td{
    padding: var(--size-sm);
    display: table-cell;
    gap: 0px;
    border-bottom: 0.05rem solid var(--darkGrayishBlue300);
}
.table-responsive{
  overflow-x: scroll;
}

/*  CONTEXT MENU    */

.collection {
    padding-left: 0;
    list-style-type: none;
    margin: .5rem 0 1rem 0;
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    overflow: hidden;
    position: relative;
  }
  
  .collection-item {
    background-color: #fff;
    line-height: 1.5rem;
    padding: 10px 20px;
    margin: 0;
    border-bottom: 1px solid #e0e0e0;
    list-style-type: none;
  }
  
  
  
  
  
  /* Animatinons */
  @-webkit-keyframes fadeIn {
    from {
      opacity: 0;
    }
    to {
      opacity: 1.0;
    }
  }
  
  @keyframes fadeIn {
    from {
      opacity: 0;
    }
    to {
      opacity: 1.0;
    }
  }
  
  @-webkit-keyframes fadeOut {
    from {
      opacity: 1.0;
    }
    to {
      opacity: 0.0;
    }
  }
  
  @keyframes fadeOut {
    from {
      opacity: 1.0;
    }
    to {
      opacity: 0.0;
    }
  }
  
  .is-fadingIn {
    -webkit-animation: fadeIn 0.1s ease-out;
    animation: fadeIn 0.1s ease-out;
    opacity: 1.0;
    display: block;
  }
  
  .is-fadingOut {
    -webkit-animation: fadeOut 0.1s ease-out;
    animation: fadeOut 0.1s ease-out;
    opacity: 0.0;
    display: block;
  }
