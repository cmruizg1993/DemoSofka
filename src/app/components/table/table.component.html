<div class="container-fluid bg-white table-responsive">
    <table class="bg-white mt-2">
        <thead class="" *ngIf="headers.length > 0">
            <th *ngFor="let h of headers">
                <div *ngIf="!h.hidden " class="tooltip d-flex">{{h.label}}
                    <span *ngIf="h.info" class="d-flex" style="align-items: end;">
                        <img src="assets/img/info.svg" width="14px" style="margin-left: 0.5rem;">                        
                    </span>
                    <span *ngIf="h.info" class="tooltiptext">{{h.info}}</span>
                </div>
            </th>
            <th *ngIf="context"></th>
        </thead>
        <tbody *ngIf="data.length > 0" class="collection">
            <tr *ngFor="let item of view; index as i" >
                <td *ngFor="let h of headers">
                    <span *ngIf="!h.hidden && !h.isImage">{{getValue(i, h )}}</span>
                    <img *ngIf="h.isImage" [src]="getValue(i, h)" width="50px" alt="img" style="border-radius: 50%;">
                </td>
                <td *ngIf="context">
                    <button class="btn-icon" (click)="openContextMenu($event, item); $event.preventDefault();">
                        <img src="assets/img/dots.svg" width="14px">
                    </button>
                </td>
            </tr>
        </tbody>
    </table>
    <div class="pager bg-white mt-1">
        <app-pager [dataLength]="data.length" (pagerUpdate)="pagerUpdate($event)"></app-pager>   
    </div>
</div>
<app-context-menu *ngIf="context" 
    [contextMenu]="context"
    [showContextMenu$]="showContextMenu$.asObservable()"
    >
</app-context-menu>