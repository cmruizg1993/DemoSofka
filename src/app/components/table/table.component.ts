import { Component, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { IContextMenu } from '../../interfaces/i-context-menu';
import { IEventContextMenu } from '../../interfaces/i-event-context-menu';

@Component({
  selector: 'app-table',
  templateUrl: './table.component.html',
  styleUrl: './table.component.css'
})
export class TableComponent {
  @Input() data: any[] = [];
  @Input() headers: any[] = [];
  @Input() context?: IContextMenu ;
  page = 1;
  pageSize = 5;
  view: any = [];

  showContextMenu$ = new Subject<IEventContextMenu>();

    
  ngAfterContentChecked(): void {
    //Called after every check of the component's or directive's content.
    //Add 'implements AfterContentChecked' to the class.
    this.tableRefresh();
  }
  getValue(index: number, header: any){
    let value = this.view[index][`${header.key}`];
    const isImage = header.isImage;
    return isImage ? `assets/img/${value}`: value
  }

  pagerUpdate(pagerProperties: any){
    this.pageSize = pagerProperties.pageSize;
    this.page = pagerProperties.page;
    this.tableRefresh();
  }

  tableRefresh(){
    const start = this.pageSize*(this.page - 1);
    const end = start + this.pageSize
    this.view = this.data.slice(start, end);
  }
  openContextMenu(event: MouseEvent, item: any){
    if(this.context){
      this.context.data = item;
      this.showContextMenu$.next(
        {
          event: event,
          data: item
        }
      )
    }
  }

}
