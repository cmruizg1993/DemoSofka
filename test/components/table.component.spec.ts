import { ComponentFixture, TestBed } from '@angular/core/testing';
import { TableComponent } from '../../src/app/components/table/table.component';
import { RouterTestingModule } from '@angular/router/testing';
import { PagerComponent } from '../../src/app/components/pager/pager.component';
import { SelectComponent } from '../../src/app/components/select/select.component';
import { NO_ERRORS_SCHEMA } from '@angular/core';



describe('TableComponent', () => {
  let fixture: ComponentFixture<TableComponent>;
  let compiled: HTMLElement;
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      declarations: [
        TableComponent,
      ],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
    fixture = TestBed.createComponent(TableComponent);
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('Se debe crear el componente', () => {
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });
  it(`Debe coincidir con el snapshot'`, () => {
    expect(compiled).toMatchSnapshot();
  });
})