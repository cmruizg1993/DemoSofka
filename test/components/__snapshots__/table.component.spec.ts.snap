// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent Debe coincidir con el snapshot' 1`] = `
<div
  id="root1"
>
  <div
    class="container-fluid bg-white table-responsive"
  >
    <table
      class="bg-white mt-2"
    >
      
      
    </table>
    <div
      class="pager bg-white mt-1"
    >
      <app-pager />
    </div>
  </div>
  
</div>
`;
